File: java/src/main/java/pro/wata/sdk/digitalgoods/stars/StarsClient.java
package pro.wata.sdk.digitalgoods.stars;

import pro.wata.sdk.http.HttpTransport;

import java.util.Map;

/**
 * Telegram Stars (токен терминала Stars — всегда отдельный от Steam, см. п.1
 * SPEC.md). Количество звёзд: 50–50000.
 */
public final class StarsClient {

    private final HttpTransport transport;

    public StarsClient(HttpTransport transport) {
        this.transport = transport;
    }

    /** {@code GET /stars/price}. */
    public StarsPrice price() {
        return transport.get("/stars/price", null, StarsPrice.class);
    }

    /**
     * Создаёт заказ ({@code POST /stars}). SPEC.md не называет имя поля для
     * количества звёзд — передайте тело целиком (например {@code count},
     * получателя и т.д.). Если сумма заказа превышает порог
     * автоподтверждения, ответ придёт со статусом {@link pro.wata.sdk.model.StarsOrderStatus#REVIEW}
     * и заказ не будет выполнен, пока его не подтвердят.
     *
     * <p>Изменяющий запрос — по умолчанию не повторяется при сбое.
     */
    public StarsOrder createOrder(Map<String, Object> body) {
        return transport.post("/stars", body, StarsOrder.class, false);
    }

    /** {@code GET /stars/order/{id}}. */
    public StarsOrder getOrder(String id) {
        return transport.get("/stars/order/" + id, null, StarsOrder.class);
    }

    /** Подтверждает заказ в статусе {@code Review} ({@code POST /stars/order/{id}/confirm}). Не повторяется. */
    public StarsOrder confirmOrder(String id) {
        return transport.post("/stars/order/" + id + "/confirm", Map.of(), StarsOrder.class, false);
    }

    /** Отклоняет заказ в статусе {@code Review} ({@code POST /stars/order/{id}/reject}). Не повторяется. */
    public StarsOrder rejectOrder(String id) {
        return transport.post("/stars/order/" + id + "/reject", Map.of(), StarsOrder.class, false);
    }
}

File: dotnet/src/Wata.Sdk/DigitalGoods/Stars/StarsClient.cs
using Wata.Sdk.Errors;
using Wata.Sdk.Http;

namespace Wata.Sdk.DigitalGoods.Stars;

/// <summary>
/// Telegram Stars. Требует WataOptions.StarsToken — терминал Stars всегда отдельный
/// от терминала Steam. Заказы дороже порога уходят в статус Review и не исполняются
/// автоматически: их нужно явно подтвердить или отклонить.
/// </summary>
public sealed class StarsClient
{
    private readonly WataHttpTransport? _transport;

    internal StarsClient(WataHttpTransport? transport) => _transport = transport;

    private WataHttpTransport Transport => _transport ?? throw new WataConfigurationException(
        "Для работы с Telegram Stars не задан токен терминала. Передайте WataOptions.StarsToken при создании WataClient. " +
        "Обратите внимание: терминал Stars не может совпадать с терминалом Steam.");

    /// <summary>GET /stars/price — стоимость и допустимый диапазон количества звёзд.</summary>
    public Task<StarsPrice> GetPriceAsync(StarsPriceRequest request, CancellationToken cancellationToken = default)
    {
        var query = new QueryStringBuilder()
            .Add("StarsCount", request.StarsCount)
            .Add("Currency", request.Currency)
            .ToString();

        return Transport.SendAsync<StarsPrice>(HttpMethod.Get, "/api/stars/price", query, null, requiresAuth: true, allowRetry: true, cancellationToken);
    }

    /// <summary>POST /stars — создать заказ. Изменяющий запрос — не повторяется при сбое.</summary>
    public Task<StarsOrder> CreateOrderAsync(CreateStarsOrderRequest request, CancellationToken cancellationToken = default) =>
        Transport.SendAsync<StarsOrder>(HttpMethod.Post, "/api/stars", null, request, requiresAuth: true, allowRetry: false, cancellationToken);

    /// <summary>GET /stars/order/{id} — статус заказа.</summary>
    public Task<StarsOrder> GetOrderAsync(string id, CancellationToken cancellationToken = default) =>
        Transport.SendAsync<StarsOrder>(HttpMethod.Get, $"/api/stars/order/{Uri.EscapeDataString(id)}", null, null, requiresAuth: true, allowRetry: true, cancellationToken);

    /// <summary>
    /// POST /stars/order/{id}/confirm — подтвердить заказ в статусе Review. Изменяющий
    /// запрос — не повторяется при сбое.
    /// </summary>
    public Task<StarsOrder> ConfirmOrderAsync(string id, CancellationToken cancellationToken = default) =>
        Transport.SendAsync<StarsOrder>(HttpMethod.Post, $"/api/stars/order/{Uri.EscapeDataString(id)}/confirm", null, null, requiresAuth: true, allowRetry: false, cancellationToken);

    /// <summary>
    /// POST /stars/order/{id}/reject — отклонить заказ в статусе Review. Изменяющий
    /// запрос — не повторяется при сбое.
    /// </summary>
    public Task<StarsOrder> RejectOrderAsync(string id, CancellationToken cancellationToken = default) =>
        Transport.SendAsync<StarsOrder>(HttpMethod.Post, $"/api/stars/order/{Uri.EscapeDataString(id)}/reject", null, null, requiresAuth: true, allowRetry: false, cancellationToken);
}
